packages/typescript-estree/src/create-program/createProjectService.ts

```diff
@@ -46,5 +58,16 @@ export function createProjectService(): TypeScriptProjectService {
     },
     session: undefined,
   });
+
+  return {
+    allowDefaultProjectForFiles: new Set(
+      typeof options === 'object'
+        ? options.allowDefaultProjectForFiles
+            ?.flatMap(pattern => globby.sync(pattern))
```

Comment:
hmmmm I wonder if we could do this check after-the-fact.
EG like store the glob(s) themselves and then if a file fails to be found - then compare the glob with the file path to look for a match.

that way we never ever resolve the globs to disk paths - saving time and ensuring we don't fall out-of-sync with disk state.

one weakness of the current approach is that new files (eg if you add a new `.eslintrc.js` in your codebase) would fail the IDE match and the user would need to restart their IDE to get it working.

Reply:
I like the lazy-loading idea for the file-not-found case a lot, but we also need the resolved globs in the case of the file-found case:

```ts
if (allowDefaultProjectForFiles.has(filePath)) {
  throw new Error(
    `${filePath} was included by allowDefaultProjectForFiles but also was found in the project service. Consider removing it from allowDefaultProjectForFiles.`,
  );
}
```

Given we already break down in the case of files changing, is there much benefit to lazy-globbing here?

Reply:
do we need to resolve them, though?
Shouldn't we be able to convert that check instead to
```
if (file matches globs) {
  throw new Error();
}
```

a library like minimatch would allow us to handle this without any disk ops

Reply:
Hmm, I was thinking there may be edge cases around files existing on disk but not being included in the TSConfig - or some inverted version of that... But that doesn't seem very likely. And I suppose we can cross that bridge if we get to it.

Reply:
Amusingly, `minimatch` was already many times over a sub-dependency. No significant change to the lockfile.